skate-plugin/src/main/kotlin/com/slack/sgp/intellij/featureflags/FeatureFlagExtractor.kt

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2023 Slack Technologies, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.slack.sgp.intellij.featureflags
+
+import com.intellij.openapi.diagnostic.Logger
+
+/**
+ * Responsible for extracting feature flags. Searches for enum entries annotated with 'FeatureFlag'
+ * to identify feature flags.
+ */
+class FeatureFlagExtractor {
+
+  private val log: Logger = Logger.getInstance(FeatureFlagExtractor::class.java)
+
+  // Caches the feature flags for a given PSI file to optimize repeated lookups
+  companion object {
+    private val featureFlagCache = mutableMapOf<com.intellij.psi.PsiFile, List<String>>()
+  }
+
+  fun setFeatureFlagsForPsiFile(psiFile: com.intellij.psi.PsiFile, flags: List<String>) {
+    featureFlagCache[psiFile] = flags
+  }
+
+  fun getFeatureFlagsForPsiFile(psiFile: com.intellij.psi.PsiFile): List<String>? {
+    return featureFlagCache[psiFile]
+  }
+  /**
+   * Extracts the names of feature flags from the provided PSI file. Only processes Kotlin files.
+   *
+   * @param psiFile The PSI representation of the file to process.
+   * @return A list of feature flag names in a file
+   */
+  fun extractFeatureFlags(psiFile: com.intellij.psi.PsiFile): List<String> {
+    log.info("Looking for feature flags in a file: ${psiFile.toString()}")
+    val enumsWithAnnotation = findAnnotatedEnums(psiFile)
+    log.info("Found feature flags: $enumsWithAnnotation")
+    return enumsWithAnnotation
+  }
+
+  /** Recursively searches the given PSI file for enums with 'FeatureFlag' annotations. */
+  private fun findAnnotatedEnums(psiFile: Any): List<String> {
+    val annotatedEnumEntries = mutableListOf<String>()
+
+    fun addIfAnnotatedEnum(element: Any) {
+      if (isKtEnumEntry(element) && hasFeatureFlagAnnotation(element)) {
+        element.javaClass
+          .getMethod("getName")
+          .invoke(element)
+          .takeIf { it is String }
+          ?.let { annotatedEnumEntries.add(it as String) }
```

Comment:
This isn't quite correct. The key logic will use the key defined in the `@FeatureFlag` annotation if it's not blank, and otherwise fallback to the lowercase (US locale) entry name as a default.

Reply:
No longer relevant

Reply:
I think it is, the updated code still does not account for a `key` value

Reply:
Being handled [here](https://github.com/slackhq/slack-gradle-plugin/pull/576/commits/3d793946ee4cb54b388a873d84ce7576699e5376#diff-0073d4ea07ce931d766fbad1989d075c4289e50411408234e22459bde49a29bdR45)